content/blog/2019-09-25-making-function-calls-work.md

```diff
@@ -0,0 +1,299 @@
++++
+title = "Bril()"
+extra.author = "Gautam Mekkat, Eashan Garg, and Sameer Lal"
+extra.bio = """
+"""
++++
+
+## Bril()
+
+### Goal
+Our goal for this project was to make function calls work. Functions reduce code duplication and allow programmers to abstract away common functionality. This simplifies the development and improves the readability of complex programs. To make functions even more useful, we introduce function parameters, return types, optional type annotations for parameters, nested function definitions, and a simple module system. We also offer the option to pass command-line arguments to the main function.
+
+### Design
+
+#### Adding support for function calls
+The provided interpreter already supports the evaluation of functions. All we have to do is add a *call* instruction to the language grammar and link that to the interpreter. To limit function scope, we create a new environment (a map from variable names to values) every time a function is called.
+
+Below is a Bril program that demonstrates this functionality.
+
+    # This program prints out 100 and exits. Nothing too exciting here.
+    main {
+      call func;
+    }
+    
+    func {
+      v0: int = const 100;
+      print v0;
+    }
+
+Since the interpreter scans function definitions before executing anything, functions can be defined in any order. The above program can be rewritten as:
+
+    func {
+      v0: int = const 100;
+      print v0;
+    }
+
+    main {
+      call func;
+    }
+
+
+#### Adding function parameters
+To add support for function parameters, we first update the grammars for function definitions and calls to take whitespace-delimited lists of variable names. The interpreter's `call` operation handler first extracts the values for all arguments from the current environment map. It then pre-populates the callee's environment with the function parameters mapped to these values. This new environment is then used when evaluating the called function.
+
+Below is a Bril program that demonstrates the use of function parameters.
+
+    # This program also prints out 100 and exits.
+    main {
+      call print_double 50;
+    }
+    
+    print_double x {
```

Comment:
It's somewhat surprising that the implementation doesn't give types to arguments—all other variables have types, and so does the return type. Any particular reason you did it this way? (No need to change the design, of course—it's a perfectly reasonable design decision; I just think it would be nice to briefly mention why you did it this way.)